File: src/main/java/org/cubexmc/railway/physics/LeashCoupler.java
package org.cubexmc.railway.physics;

import java.util.ArrayList;
import java.util.List;

import org.bukkit.Location;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Minecart;
import org.bukkit.entity.Mob;
import org.bukkit.util.Vector;
import org.cubexmc.railway.Railway;
import org.cubexmc.railway.train.TrainInstance;
import org.cubexmc.railway.util.SchedulerUtil;

/**
 * Manages invisible leashable entities between cars for visual coupling.
 */
public class LeashCoupler {

    private final Railway plugin;
    private final TrainInstance train;
    private final List<LivingEntity> dummies = new ArrayList<>();
    private double offsetY;

    public LeashCoupler(Railway plugin, TrainInstance train) {
        this.plugin = plugin;
        this.train = train;
        this.offsetY = plugin.getLeashOffsetY();
    }

    public void start() {
        cleanup();
        List<Minecart> cars = train.getConsist().getCars();
        if (cars.size() < 2) return;

        EntityType type = parseEntityType(plugin.getLeashMobTypeRaw());
        if (type == null) return;

        for (int i = 1; i < cars.size(); i++) {
            Minecart prev = cars.get(i - 1);
            Minecart cur = cars.get(i);
            if (prev == null || cur == null || prev.isDead() || cur.isDead()) continue;

            Location base = midpoint(prev.getLocation(), cur.getLocation()).add(0, offsetY, 0);
            LivingEntity dummy = base.getWorld().spawn(base, type.getEntityClass().asSubclass(LivingEntity.class), ent -> {
                ent.setInvisible(true);
                ent.setSilent(true);
                ent.setInvulnerable(true);
                ent.setGravity(false);
                ent.setCollidable(false);
                if (ent instanceof Mob) {
                    ((Mob) ent).setAI(false);
                }
            });
            // Set leash holder to previous cart (visual rope between prev and dummy)
            try {
                dummy.setLeashHolder(prev);
            } catch (Throwable ignored) {}
            dummies.add(dummy);
        }
    }

    public void update() {
        List<Minecart> cars = train.getConsist().getCars();
        if (dummies.isEmpty() || cars.size() < 2) return;
        int count = Math.min(dummies.size(), Math.max(0, cars.size() - 1));
        for (int i = 0; i < count; i++) {
            LivingEntity dummy = dummies.get(i);
            Minecart prev = cars.get(i);
            Minecart cur = cars.get(i + 1);
            if (dummy == null || dummy.isDead() || prev == null || cur == null || prev.isDead() || cur.isDead()) continue;
            Location target = midpoint(prev.getLocation(), cur.getLocation()).add(0, offsetY, 0);
            SchedulerUtil.teleportEntity(dummy, target);
            // Ensure leash still tied to prev
            if (!dummy.isLeashed() || dummy.getLeashHolder() != prev) {
                try { dummy.setLeashHolder(prev); } catch (Throwable ignored) {}
            }
        }
    }

    public void cleanup() {
        for (LivingEntity le : new ArrayList<>(dummies)) {
            if (le != null && !le.isDead()) {
                try { le.setLeashHolder(null); } catch (Throwable ignored) {}
                le.remove();
            }
        }
        dummies.clear();
    }

    private static Location midpoint(Location a, Location b) {
        if (a == null) return b;
        if (b == null) return a;
        Vector va = a.toVector();
        Vector vb = b.toVector();
        Vector vm = va.clone().add(vb).multiply(0.5);
        return new Location(a.getWorld(), vm.getX(), vm.getY(), vm.getZ());
    }

    private static EntityType parseEntityType(String raw) {
        if (raw == null) return null;
        try {
            EntityType t = EntityType.valueOf(raw.trim().toUpperCase());
            if (!t.isAlive()) return null;
            return t;
        } catch (IllegalArgumentException ex) {
            return null;
        }
    }
}

File: src/main/java/org/cubexmc/metro/physics/LeashCoupler.kt
package org.cubexmc.metro.physics

import java.util.Locale
import org.bukkit.Location
import org.bukkit.entity.EntityType
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Minecart
import org.bukkit.entity.Mob
import org.bukkit.util.Vector
import org.cubexmc.metro.Metro
import org.cubexmc.metro.train.TrainInstance
import org.cubexmc.metro.util.SchedulerUtil
import kotlin.math.max
import kotlin.math.min

/** Manages invisible leashable entities between cars for visual coupling. */
open class LeashCoupler(
    private val plugin: Metro,
    private val train: TrainInstance,
) {

    private val dummies = ArrayList<LivingEntity>()
    private var offsetY = plugin.leashOffsetY

    open fun start() {
        cleanup()
        val cars = train.consist.getCars()
        if (cars.size < 2) {
            return
        }

        val type = parseEntityType(plugin.leashMobTypeRaw) ?: return

        for (index in 1 until cars.size) {
            val previous = cartOrNull(cars, index - 1)
            val current = cartOrNull(cars, index)
            if (previous == null || current == null || previous.isDead || current.isDead) {
                continue
            }

            val base = midpoint(previous.location, current.location)?.add(0.0, offsetY, 0.0)
                ?: throw NullPointerException("midpoint")
            val world = base.world ?: throw NullPointerException("world")
            val entityClass = type.entityClass?.asSubclass(LivingEntity::class.java)
                ?: throw NullPointerException("entity class")
            val dummy = world.spawn(base, entityClass) { entity ->
                entity.isInvisible = true
                entity.isSilent = true
                entity.isInvulnerable = true
                entity.setGravity(false)
                entity.isCollidable = false
                if (entity is Mob) {
                    entity.setAI(false)
                }
            }
            // Set leash holder to previous cart (visual rope between previous and dummy)
            try {
                dummy.setLeashHolder(previous)
            } catch (_: Throwable) {
                // Some server implementations reject this holder type; the visual dummy still remains harmless.
            }
            dummies.add(dummy)
        }
    }

    open fun update() {
        val cars = train.consist.getCars()
        if (dummies.isEmpty() || cars.size < 2) {
            return
        }
        val count = min(dummies.size, max(0, cars.size - 1))
        for (index in 0 until count) {
            val dummy = dummyOrNull(index)
            val previous = cartOrNull(cars, index)
            val current = cartOrNull(cars, index + 1)
            if (dummy == null || dummy.isDead || previous == null || current == null ||
                previous.isDead || current.isDead
            ) {
                continue
            }
            val target = midpoint(previous.location, current.location)?.add(0.0, offsetY, 0.0)
                ?: throw NullPointerException("midpoint")
            SchedulerUtil.teleportEntity(dummy, target)
            // Ensure leash is still tied to previous
            if (!dummy.isLeashed || dummy.leashHolder !== previous) {
                try {
                    dummy.setLeashHolder(previous)
                } catch (_: Throwable) {
                    // Keep updating the visual dummy even if the server rejects the leash operation.
                }
            }
        }
    }

    open fun cleanup() {
        for (entity in ArrayList(dummies)) {
            val leashEntity: LivingEntity? = entity
            if (leashEntity != null && !leashEntity.isDead) {
                try {
                    leashEntity.setLeashHolder(null)
                } catch (_: Throwable) {
                    // Removal below still cleans up the visual dummy.
                }
                leashEntity.remove()
            }
        }
        dummies.clear()
    }

    /** Preserve the old Java implementation's defensive null checks for list elements. */
    private fun cartOrNull(cars: List<Minecart>, index: Int): Minecart? = cars[index]

    /** Preserve the old Java implementation's defensive null check for dummy list elements. */
    private fun dummyOrNull(index: Int): LivingEntity? = dummies[index]

    companion object {
        @JvmStatic
        private fun midpoint(first: Location?, second: Location?): Location? {
            if (first == null) {
                return second
            }
            if (second == null) {
                return first
            }
            val firstVector = first.toVector()
            val secondVector = second.toVector()
            val midpoint: Vector = firstVector.clone().add(secondVector).multiply(0.5)
            return Location(first.world, midpoint.x, midpoint.y, midpoint.z)
        }

        @JvmStatic
        private fun parseEntityType(raw: String?): EntityType? {
            if (raw == null) {
                return null
            }
            return try {
                val type = EntityType.valueOf(raw.trim().uppercase(Locale.getDefault()))
                if (type.isAlive) type else null
            } catch (_: IllegalArgumentException) {
                null
            }
        }
    }
}
